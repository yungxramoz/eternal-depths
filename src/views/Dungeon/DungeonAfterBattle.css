.action-buttons-container {
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  justify-content: end;
}
