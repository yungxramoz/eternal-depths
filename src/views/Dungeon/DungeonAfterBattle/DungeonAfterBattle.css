.action-buttons-container {
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  justify-content: flex-end;
}

.reward-selection-container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 1em;
}
