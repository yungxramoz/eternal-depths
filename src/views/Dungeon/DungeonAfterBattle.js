import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import RpgButton from '../../components/atoms/RpgButton/RpgButton'
import { recoverHp } from '../../store/character/characterSlice'
import { gameWon, nextStage } from '../../store/game/gameSlice'
import './DungeonAfterBattle.css'

const DungeonAfterBattle = () => {
  const dispatch = useDispatch()
  const stage = useSelector((state) => state.game.stage)

  const localNextStage = () => {
    dispatch(nextStage())
    dispatch(recoverHp(15))
  }

  return (
    <>
      <h1>Stage {stage}</h1>
      <div className="action-buttons-container">
        <RpgButton onClick={() => dispatch(gameWon())} text="Escape" />
        <RpgButton onClick={localNextStage} text="Next Stage" />
      </div>
    </>
  )
}

export default DungeonAfterBattle
