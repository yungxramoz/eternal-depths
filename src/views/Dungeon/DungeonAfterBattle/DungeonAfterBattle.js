import React, { useMemo, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import RpgButton from '../../../components/atoms/RpgButton/RpgButton'
import Modal from '../../../components/organisms/Modal/Modal'
import { equipItem, recoverHp } from '../../../store/character/characterSlice'
import { gameWon, nextStage } from '../../../store/game/gameSlice'
import { generateItem } from '../../../utils/item-generator'
import './DungeonAfterBattle.css'
import RewardButton from '../../../components/organisms/RewardButton/RewardButton'
import { RPGUI_ICON } from '../../../constants/rpgui-icon'

const DungeonAfterBattle = () => {
  const dispatch = useDispatch()
  const stage = useSelector((state) => state.game.stage)
  const [gearReward] = useState(generateItem())
  const [hpReward] = useState(Math.floor(stage / 5) * 10 + 5) //TODO move to utils
  const [showReward, setShowReward] = useState(true)

  const gearIcon = useMemo(() => {
    switch (gearReward.type) {
      case 'sword':
        return RPGUI_ICON.SWORD
      case 'shield':
        return RPGUI_ICON.SHIELD
      case 'armor':
        return RPGUI_ICON.ARMOR_SLOT
      case 'helmet':
        return RPGUI_ICON.HELMET_SLOT
      case 'greaves':
        return RPGUI_ICON.SHOES_SLOT
      default:
        return RPGUI_ICON.EMPTY_SLOT
    }
  }, [gearReward.type])

  const itemStats = () => {
    return Object.entries(gearReward.stats)
      .filter(([key, value]) => value > 0)
      .map(([key, value]) => `${key.slice(0, 3).toUpperCase()}:${value}`)
      .join(', ')
  }

  const selectGear = () => {
    dispatch(equipItem(gearReward))
    setShowReward(false)
  }

  const selectHpRecovery = () => {
    dispatch(recoverHp(hpReward))
    setShowReward(false)
  }

  const localNextStage = () => {
    dispatch(nextStage())
    dispatch(recoverHp(15))
  }

  return (
    <>
      {showReward && (
        <Modal
          title="Encounter slayed!"
          dismissable={false}
          onClose={() => setShowReward(false)}
        >
          <p>Choose your reward:</p>
          <div className="reward-selection-container">
            <RewardButton
              icon={gearIcon}
              title={gearReward.name}
              subtitle={gearReward.rarity}
              description={itemStats()}
              onClick={() => selectGear()}
            ></RewardButton>

            <RewardButton
              icon={RPGUI_ICON.POTION_RED}
              title="HP Recovery"
              subtitle="Common"
              description={`Recover ${hpReward} HP`}
              onClick={() => selectHpRecovery()}
            ></RewardButton>
          </div>
        </Modal>
      )}
      <h1>Stage {stage}</h1>
      <div className="action-buttons-container">
        <RpgButton onClick={() => dispatch(gameWon())} text="Escape" />
        <RpgButton onClick={localNextStage} text="Next Stage" />
      </div>
    </>
  )
}

export default DungeonAfterBattle
