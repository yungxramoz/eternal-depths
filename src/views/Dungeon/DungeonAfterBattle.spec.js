import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import gameReducer from '../../store/game/gameSlice'
import DungeonAfterBattle from './DungeonAfterBattle'

describe('DungeonAfterBattle', () => {
  const store = configureStore({
    reducer: {
      game: gameReducer,
    },
  })

  it('renders correctly', () => {
    render(
      <Provider store={store}>
        <DungeonAfterBattle />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('DungeonAfterBattle')
  })

  it('dispatches gameWon action when "Escape" button is clicked', () => {
    store.dispatch = jest.fn()
    render(
      <Provider store={store}>
        <DungeonAfterBattle />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Escape'))
    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith({ type: 'game/gameWon' })
  })

  it('dispatches nextStage and recoverHp actions when "Next Stage" button is clicked', () => {
    store.dispatch = jest.fn()
    render(
      <Provider store={store}>
        <DungeonAfterBattle />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Next Stage'))
    expect(store.dispatch).toHaveBeenCalledTimes(2)
    expect(store.dispatch).toHaveBeenCalledWith({ type: 'game/nextStage' })
    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'character/recoverHp',
      payload: 15,
    })
  })
})
