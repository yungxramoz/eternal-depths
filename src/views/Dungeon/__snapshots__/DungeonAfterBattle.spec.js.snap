// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DungeonAfterBattle renders correctly: DungeonAfterBattle 1`] = `
<body>
  <div>
    <h1>
      Stage 
      0
    </h1>
    <div
      class="action-buttons-container"
    >
      <button
        class="rpgui-button btn"
      >
        Escape
      </button>
      <button
        class="rpgui-button btn"
      >
        Next Stage
      </button>
    </div>
  </div>
</body>
`;
